(TK-384) Initial code

KevinCorcoran opened the pull request:
This pull request contains my code from https://github.com/puppetlabs/pe-file-sync/pull/257, @cprice404's code from https://github.com/cprice404/pe-file-sync/commit/a926e3ce15a4612b0245d1c0ed3ed4dce160e19f#diff-9e25f3a6c56b4a21bd72fde5efd54313, and I think I've addressed all of the great feedback you guys gave me on the previous PR.

The only thing in here that I think should be of any surprise is 79d287147357223c03be171395f98522606837ec.  I included that because it turned out to be pretty easy to do; I'm reasonably happy with how it turned out, but it's also pretty different than any other TK services I've seen before, so I'm not dead-set on that approach.  I wouldn't mind just ripping it out if y'all don't like it - shouldn't be too hard to burn down.

Other than that, I've filed a separate ticket about the whole mess of `.poll` vs. `.take`, `future` vs. scheduler and whatnot.

I think this probably deserves at least a little more test coverage than what it has right now, especially a test case for multiple watchers, which isn't covered at all right now.  But, otherwise, I'm hoping this is close-ish to merge-able.

cprice404 committed:
(TK-384) Introduce `watcher` in service protocol

KevinCorcoran committed:
(TK-384) Add FilesystemWatchService

Add a new TK service protocol and implementation to recursively watch
directories on the filesystem for changes and invoke callbacks when they
occur.  The implementation uses java.nio.file.WatchService and is
heavily inspired by WatchDir.java from Java's official documentation at
https://docs.oracle.com/javase/tutorial/essential/io/notification.html

KevinCorcoran committed:
(maint) Remove 'enterprise' from ns/package names

KevinCorcoran committed:
(TK-384) Introduce :unknown event type

... into the service protocol and use it for OVERFLOW events.

KevinCorcoran committed:
(TK-384) Add error handling

... to ensure that all WatchServices are closed during service 'stop'.

KevinCorcoran committed:
(TK-384) Cleanup docstrings on protocol

KevinCorcoran committed:
(maint) Use Oracle JDK 8 in Travis CI

This leaves Travis CI with coverage on one OpenJDK, one Oracle JDK, one
Java 7 and one Java 8.

KevinCorcoran committed:
(TK-384) Introduce Watcher schema

KevinCorcoran committed:
(TK-384) Remove file path normalization

KevinCorcoran committed:
(TK-384) Introduce Watcher protocol

This commit introduces a Watcher protocol in the same namespace as the
FilesystemWatchService protocol.  The add-watch-dir! and add-callback!
functions are moved into the new protocol.  The protocol is implemented
by a defrecord in the core namespace which allows its state to be easily
accessed by the service implementation.

KevinCorcoran committed:
(maint) Remove ! from service function name

KevinCorcoran committed:
(maint) Add comment

KevinCorcoran committed:
(maint) Remove unconventional usages of 'testing'

KevinCorcoran committed:
(TK-384) Fix bug and add test coverage

This commit adds additional, basic test coverage for creating and
deleting a directory and nested file after the service has been started.
These test cases unearthed a bug in the registration of these new
directories, which is also fixed in this commit.

KevinCorcoran committed:
(TK-384) Add test case for directory re-creation

KevinCorcoran committed:
(TK-384) Add test and docs for watch dir deletion

This commit adds a test case for the behavior when a watched directory
is deleted.  It also updates the protocol docs to explain the behavior
in this situation.

cprice404 commented on project.clj:
```diff
@@ -22,13 +22,23 @@
-  :lein-release {:scm         :git
-                 :deploy-via  :lein-deploy}
+  :lein-release {:scm :git
+                 :deploy-via :lein-deploy}
```
I think we actually don't need this boilerplate anymore; it was for the old lein-release plugin, and {{release}} is a task built in to lein itself now.  I think you can also get rid of the reference to the plugin above (and in any other repos you see this stuff still hanging around in).

cprice404 commented:
Travis failures seem like they might be a real thing, guessing another difference between osx-- and linux.

cprice404 commented:
+1 other than travis failures, your API change seems ~legit to me

cprice404 commented:
oh, and yeah, a test or two with multiple watchers would be good, but I wouldn't necessarily block the PR on that if it's holding up other tickets.

KevinCorcoran committed:
(TK-384) Docs/test update for watched dir deletion

This commit updates the protocol documentation to more correctly state
the behavior when a watched directory is deleted - namely, it is
unspecified by the API, and platform-dependent in the current
implementation.

As a result, the test case for this is deleted, since the behavior of
the current implementation is so different between OSX and Linux that
it can't be tested in a sane, platform-agnostic way.

KevinCorcoran committed:
(maint) Delete lein-release config in project.clj

It's no longer necessary now that leiningen has a built-in release task.

KevinCorcoran commented:
@cprice404 Thanks for the review, I pushed 2 commits which I think address your feedback.  I'm going to keep working on some more test coverage but you I agree with your point that this could probably be merged without the extra tests in the interest of un-blocking other work.

KevinCorcoran committed:
(TK-384) Add tests for multi watch dirs / watchers

This commit adds two new test cases - one for creating and using
multiple Watchers, the other for watching multiple directories on a
single Watcher.

KevinCorcoran commented:
I added one more commit with the extra test coverage.  I'm hoping you and @camlow325 find it to be sufficient ... this stuff seems like I could write tests for it until the cows come home and I'd still be able to come up with more tests to write.

KevinCorcoran committed:
(maint) Add Development docs to README.md

haus commented on src/clj/puppetlabs/services/watcher/filesystem_watch_service.clj:
```diff
@@ -0,0 +1,45 @@
+(ns puppetlabs.services.watcher.filesystem-watch-service
```
This looks inconsistent with other tk services. looks like they mostly start with
`puppetlabs.trapperkeeper.services.` and not `puppetlabs.services.`

KevinCorcoran replied:
Oh.  Interesting.  To start, I followed the pattern in file sync, and then remove an `enterprise` level when I moved the code into this repo.

Kinda weird, though, to have things in `puppetlabs.trapperkeeper` which aren't part of Trapperkeeper itself.  But I guess that ship has sailed.

haus commented on src/clj/puppetlabs/services/protocols/filesystem_watch_service.clj:
```diff
@@ -0,0 +1,38 @@
+(ns puppetlabs.services.protocols.filesystem-watch-service
```
Same as a comment in the implementation, but other tk service protocols use the `puppetlabs.trapperkeeper.services.protocols.` namespace. Seems like consistency would be good there.

camlow325 commented on test/puppetlabs/services/watcher/filesystem_watch_service_test.clj:
```diff
@@ -0,0 +1,387 @@
+           (let [events #{{:path sub-dir
+                           :type :delete}}]
+             (is (= events (wait-for-events results events)))))
+         (testing "Re-creating the directory fires an event as expcted"
```
s/expcted/expected

camlow325 commented:
👍 from me.  I didn't have a chance to review all of the tests yet but what I did looked good to me.  I'm good with this being merged when @haus feedback has been addressed.

KevinCorcoran committed:
(maint) Fix typo

KevinCorcoran committed:
(TK-384) Rename namespaces per convention

The convention for namespace prefixes in other open-source Trapperkeeper
projects is "puppetlabs.trapperkeeper.services" not
"puppetlabs.services".

KevinCorcoran commented:
@cprice404 @camlow325 @haus I think we're pretty much at consensus on this (I recently added 2 more commits to address the last bits of feedback).  I'm going to self-merge this so I can do an initial release to Clojars, but I'm happy to address any other issues you should find in a subsequent PR.
